spring-rabbit/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitListenerAnnotationBeanPostProcessor.java

```diff
@@ -487,61 +487,81 @@ else if (resolvedValueToUse instanceof Iterable) {
-							resolveArguments(bindingExchange.arguments()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
-							resolveArguments(binding.arguments()));
-				}
-				else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
-					exchange = new HeadersExchange(exchangeName,
-							resolveExpressionAsBoolean(bindingExchange.durable()),
-							resolveExpressionAsBoolean(bindingExchange.autoDelete()),
-							resolveArguments(bindingExchange.arguments()));
-					actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, resolvedKey,
-							resolveArguments(binding.arguments()));
-				}
-				else {
-					throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
-				}
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
-						exchange);
-				((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
-						actualBinding);
+				declareExchangeAndBinding(binding, queueName, bindingExchange);
 			}
 		}
 		return queues.toArray(new String[queues.size()]);
 	}
 
+	private void declareExchangeAndBinding(QueueBinding binding, String queueName,
+			org.springframework.amqp.rabbit.annotation.Exchange bindingExchange) {
+		String exchangeName = resolveExpressionAsString(bindingExchange.value(), "@Exchange.exchange");
+		String exchangeType = resolveExpressionAsString(bindingExchange.type(), "@Exchange.type");
+		String routingKey = resolveExpressionAsString(binding.key(), "@QueueBinding.key");
+		Exchange exchange;
+		Binding actualBinding;
+		if (exchangeType.equals(ExchangeTypes.DIRECT)) {
+			exchange = directExchange(bindingExchange, exchangeName);
+			actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, routingKey,
+					resolveArguments(binding.arguments()));
+		}
+		else if (exchangeType.equals(ExchangeTypes.FANOUT)) {
+			exchange = fanoutExchange(bindingExchange, exchangeName);
+			actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, "",
+					resolveArguments(binding.arguments()));
+		}
+		else if (exchangeType.equals(ExchangeTypes.TOPIC)) {
+			exchange = topicExchange(bindingExchange, exchangeName);
+			actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, routingKey,
+					resolveArguments(binding.arguments()));
+		}
+		else if (exchangeType.equals(ExchangeTypes.HEADERS)) {
+			exchange = headersExchange(bindingExchange, exchangeName);
+			actualBinding = new Binding(queueName, DestinationType.QUEUE, exchangeName, routingKey,
+					resolveArguments(binding.arguments()));
+		}
+		else {
+			throw new BeanInitializationException("Unexpected exchange type: " + exchangeType);
+		}
+		((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+				exchange);
+		((ConfigurableBeanFactory) this.beanFactory).registerSingleton(exchangeName + ++this.increment,
+				actualBinding);
```

Comment:
H-m. The same `exchangeName` for `Binding` bean, even if it is just a prefix.
How about to have `queueName/exchangeName` combination for the `Binding` bean?
I know it is an existing code, but I won't mind to discuss and fix it here.